crypto/x509/x509_req.c

```diff
@@ -286,6 +286,15 @@ void X509_REQ_get0_signature(const X509_REQ *req, const ASN1_BIT_STRING **psig,
         *palg = &req->sig_alg;
 }
 
+void X509_REQ_set0_signature(X509_REQ *req, ASN1_BIT_STRING *psig,
```

Comment:
I think the memory management semantics of this function is very confusing. It took me several minutes of staring at this PR before I realised this - so I think users will also be confused. A "set0" function in OpenSSL is usually used to mean that the memory management of the passed arguments is transferred to the object. However this is only the case for *one* of these values. So one argument has its ownership transferred to the object, while the ownership remains with the caller for the other one.

Perhaps this function would be better split into two separate setters, i.e. a "set0" and a "set1"

Reply:
I took earlier advise and tried to keep close to the get0 function. Happy to split it in two functions - that also makes it perhaps easier to explain the copy of the struct v.s.  pointer.

Does that also mean the get0 needs to be changed in a get0/1-- that is sort of an API change ?

Reply:
> I took earlier advise and tried to keep close to the get0 function. Happy to split it in two functions - that also makes it perhaps easier to explain the copy of the struct v.s. pointer.

I do think it should be split.

> Does that also mean the get0 needs to be changed in a get0/1-- that is sort of an API change

Well, no the get0 is actually correct. In both cases you get a pointer to the internal structure back. There is no "copying" going on. So the memory semantics is the same for both fields - the ownership remains with the object (which is the intended meaning for get0).

Reply:
First of all: thanks @mattcaswell for spotting the memory issues. I was too much distracted by other things to notice them.


IMHO splitting the setter into two is not really optimal because signature and signature algorithm belong together. An alternative to splitting would be to make the last parameter a `const X509_ALGOR *palg`, to emphasize that the struct is passed "by reference" only for efficiency reasons, but effectively only the value is copied.

Reply:
Ok  so what is the consensus of this community  ? Happy to adapt.

**1.**  one get0 for both -- as they are a pair; or a get0 / 1 for each ?

**2.** And should the set0 be:

   **2.A:** much like all the other set0's which will do a free and the update the pointer; but not DUP (transfer mem mngt)

```
void NAMING_AUTHORITY_set0_authorityId(NAMING_AUTHORITY *n, ASN1_OBJECT* id)
{
    ASN1_OBJECT_free(n->namingAuthorityId);
    n->namingAuthorityId = id;
}
```

   **2.B:** or much like X509_NAME_set() and most other 'non 0/1'  plain XX_set()s which mostly do a DUP ?

```
int X509_NAME_set(X509_NAME **xn, const X509_NAME *name)
{
    X509_NAME *name_copy;

    if (*xn == name)
        return *xn != NULL;
    if ((name_copy = X509_NAME_dup(name)) == NULL)
        return 0;
    X509_NAME_free(*xn);
    *xn = name_copy;
    return 1;
}
```

**3)** Happy to add the other two methods - now or in a separate pull. Just let me know what is required/process wise desired.

With kind regards,

Dw.